File: web/js/reference_resolution.js
// Type-homogeneous Bridge output groups. Mirrored in
// server/reference_resolution.py.
export const REFERENCE_OUTPUT_NAMES = [
    "image_slots",
    "audio_slots",
    "reference_prompt",
    "reference_names",
];

// A recipe without a `live_outputs` declaration (detached/custom) keeps every
// output live; anything else is dead and emits its documented fallback.
export function referenceLiveOutputs(hard) {
    const declared = hard?.live_outputs;
    if (!Array.isArray(declared)) return new Set(REFERENCE_OUTPUT_NAMES);
    return new Set(declared.filter((name) => REFERENCE_OUTPUT_NAMES.includes(name)));
}

/**
 * The prompt a staged item contributes, mirroring `_assemble_prompt` in
 * `nodes/reference_core.py`. The panel shows this so the user can read and copy
 * what actually reaches the model before deciding to override it.
 *
 * `members` are pre-resolved `{name, prompt}` pairs in staged slot order.
 */
/**
 * Expand one member's slice of the derived prompt. Mirrors
 * `member_prompt_fragment` in nodes/reference_core.py.
 *
 * A pattern may use `{n}` (1-based), `{index}` (0-based), `{prompt}` and
 * `{name}` as many times as it likes. With no `{prompt}`/`{name}` placeholder
 * the member text is appended after the expanded pattern, preserving the
 * original `token: label` behaviour.
 */
export const PROJECT_SCOPED_PROMPT_TOKENS = Object.freeze([
    "{subject_n}", "{picture_n}", "{audio_n}", "{speaker_n}",
]);

export function memberPromptFragment(pattern, index, prompt, name, registryNumbers = null) {
    const memberPrompt = String(prompt || "").trim();
    const entityName = String(name || "").trim();
    const label = memberPrompt || entityName;
    if (!pattern) return label;
    const numbers = registryNumbers && typeof registryNumbers === "object" ? registryNumbers : {};
    // split/join, not String.replace: a string-literal replace substitutes only
    // the first occurrence, while Python's str.replace substitutes every one.
    // `{n}` stays lane-local; the four project-scoped tokens come from the
    // cross-lane registry so one entity numbers the same on every lane.
    const expand = (promptValue) => String(pattern)
        .split("{subject_n}").join(String(numbers.subject_n || 0))
        .split("{picture_n}").join(String(numbers.picture_n || 0))
        .split("{audio_n}").join(String(numbers.audio_n || 0))
        .split("{speaker_n}").join(String(numbers.speaker_n || 0))
        .split("{n}").join(String(index + 1))
        .split("{index}").join(String(index))
        .split("{prompt}").join(promptValue)
        .split("{name}").join(entityName);
    if (pattern.includes("{prompt}") || pattern.includes("{name}")) {
        // An empty member prompt inside a `{prompt}` pattern would otherwise
        // leave a dangling clause; fall back to the entity name rather than
        // dropping the fragment, which would hide a member whose image still
        // reaches the model.
        const useName = !memberPrompt && entityName && pattern.includes("{prompt}");
        return expand(useName ? entityName : memberPrompt).trim();
    }
    return label ? `${expand(memberPrompt)}: ${label}` : expand(memberPrompt);
}

export function deriveReferencePrompt({ promptOverride = "", members = [], soft = {} } = {}) {
    const override = String(promptOverride || "").trim();
    if (override) return override;
    const tokenPattern = String(soft?.prompt_tokens || "");
    const values = [];
    (Array.isArray(members) ? members : []).forEach((entry, index) => {
        const fragment = memberPromptFragment(tokenPattern, index, entry?.prompt, entry?.name);
        if (fragment) values.push(fragment);
    });
    const prefix = String(soft?.prompt_prefix || "").trim();
    const joined = values.join(", ");
    return [prefix, joined].filter(Boolean).join(" ");
}

function integer(value, fallback = 0) {
    const number = Number(value);
    return Number.isFinite(number) ? Math.trunc(number) : fallback;
}

/** Why a staged item did or did not reach the model for a window. */
export const REFERENCE_VERDICT = Object.freeze({
    WINNER: "winner",
    SUPERSEDED: "superseded",
    BELOW_THRESHOLD: "below_threshold",
    OUTSIDE: "outside",
    EXCLUDED: "excluded",
});

/** Display labels, shared by the timeline marks and the lane panel chips. */
export const REFERENCE_VERDICT_LABEL = Object.freeze({
    [REFERENCE_VERDICT.WINNER]: "In window",
    [REFERENCE_VERDICT.SUPERSEDED]: "Superseded",
    [REFERENCE_VERDICT.BELOW_THRESHOLD]: "Below threshold",
    [REFERENCE_VERDICT.OUTSIDE]: "Outside window",
    [REFERENCE_VERDICT.EXCLUDED]: "Excluded",
});

/**
 * Score every staged item for a window and say what happened to each one.
 *
 * This is the scoring core. `resolveEffectiveReferences` derives its winners
 * from it, so the shared JS-vs-Python winner parity test also guards this: any
 * drift here changes the winners and fails that comparison.
 *
 * `frameThresholdPct` drops an item whose in-window overlap is under that
 * percentage of its OWN span, before the survivors are scored. Unlike the
 * prompt boundary threshold there is no never-empty guard: a lane may resolve
 * to nothing, which is what makes a reference stop applying outside its scope.
 *
 * @returns {{winners: Array, verdicts: Map<number, string>}} verdicts keyed by
 *   the item's index in `referenceItems`.
 */
export function resolveReferenceVerdicts({
    referenceItems = [],
    laneCount = 1,
    sceneDuration = 0,
    windowStart = 0,
    windowEnd = sceneDuration,
    laneConfigs = [],
    frameThresholdPct = 0,
} = {}) {
    const count = Math.max(1, integer(laneCount, 1));
    const rawThreshold = Number(frameThresholdPct);
    const threshold = (Number.isFinite(rawThreshold) ? Math.max(0, Math.min(100, rawThreshold)) : 0) / 100;
    const duration = Math.max(0, integer(sceneDuration, 0));
    const start = Math.min(duration, Math.max(0, integer(windowStart, 0)));
    const end = Math.min(duration, Math.max(start, integer(windowEnd, duration)));
    const items = Array.isArray(referenceItems) ? referenceItems : [];
    const configs = Array.isArray(laneConfigs) ? laneConfigs : [];
    const winners = Array(count).fill(null);
    const scores = Array(count).fill(null);
    const verdicts = new Map();

    items.forEach((item, itemIndex) => {
        const laneIndex = integer(item?.lane_index, 0);
        if (laneIndex < 0 || laneIndex >= count || item?.muted || configs[laneIndex]?.hidden) {
            verdicts.set(itemIndex, REFERENCE_VERDICT.EXCLUDED);
            return;
        }
        const itemStart = Math.min(
            Math.max(0, integer(item?.start_frame, 0)),
            Math.max(0, duration - 1),
        );
        const rawEnd = integer(item?.end_frame, -1);
        const itemEnd = rawEnd < 0 ? duration : Math.min(duration, rawEnd);
        if (itemEnd <= itemStart) {
            verdicts.set(itemIndex, REFERENCE_VERDICT.OUTSIDE);
            return;
        }
        const overlap = Math.max(0, Math.min(itemEnd, end) - Math.max(itemStart, start));
        if (overlap <= 0) {
            verdicts.set(itemIndex, REFERENCE_VERDICT.OUTSIDE);
            return;
        }
        const coverage = overlap / (itemEnd - itemStart);
        if (threshold > 0 && coverage < threshold) {
            verdicts.set(itemIndex, REFERENCE_VERDICT.BELOW_THRESHOLD);
            return;
        }
        // A candidate that clears the threshold is provisionally the winner and
        // demotes whoever held the lane; most-specific-wins is resolved by the
        // scan order, so the loser's verdict is only final once it completes.
        const score = [coverage, itemStart, itemIndex];
        const previous = scores[laneIndex];
        const wins = !previous
            || score[0] > previous[0]
            || (score[0] === previous[0] && score[1] > previous[1])
            || (score[0] === previous[0] && score[1] === previous[1] && score[2] > previous[2]);
        if (wins) {
            if (winners[laneIndex]) verdicts.set(winners[laneIndex].itemIndex, REFERENCE_VERDICT.SUPERSEDED);
            scores[laneIndex] = score;
            winners[laneIndex] = { laneIndex, item, itemIndex };
            verdicts.set(itemIndex, REFERENCE_VERDICT.WINNER);
        } else {
            verdicts.set(itemIndex, REFERENCE_VERDICT.SUPERSEDED);
        }
    });
    return { winners, verdicts };
}

/**
 * The lane an item belongs to, normalized exactly as the scorer does.
 *
 * Export rather than re-spell: `(item.lane_index || 0)` and a raw read both
 * disagree with the scorer for `undefined`, `"2"` and `1.7`, which the scorer
 * reads as lanes 0, 2 and 1. Any caller bucketing items by lane must agree with
 * the function that decided the verdicts, or a resolved lane reads unresolved.
 */
export function referenceLaneIndex(item) {
    return integer(item?.lane_index, 0);
}

/** Why a lane failed to resolve for one chunk, most actionable first. */
export const REFERENCE_LANE_CAUSE = Object.freeze({
    BELOW_THRESHOLD: "below_threshold",
    OUTSIDE: "outside",
    EXCLUDED: "excluded",
});

/**
 * Per-lane outcome across a batch's chunks, with the CAUSE of each failure.
 *
 * `resolveEffectiveReferences` answers "did this lane resolve" and throws the
 * reason away, so a caller reading it cannot tell a deliberately scoped item
 * (never overlapped this chunk) from one the frame threshold dropped. Both flip
 * `has_reference` 1->0 and both matter, but they need different words and a
 * different urgency, and only one of them is fixed by touching a setting.
 *
 * Cause precedence per chunk: BELOW_THRESHOLD dominates, because it is the only
 * one where the item DID overlap the chunk and a setting is what dropped it.
 * EXCLUDED (muted item / hidden lane) is window-independent, so it can only ever
 * apply to every chunk at once — it never produces a flip.
 *
 * @returns {Array<{laneIndex, staged, resolved, causeCounts, dominantCause}>}
 */
export function classifyReferenceChunks(chunks = [], options = {}) {
    const { referenceItems = [], laneCount = 1 } = options;
    const count = Math.max(1, integer(laneCount, 1));
    const items = Array.isArray(referenceItems) ? referenceItems : [];
    const windows = Array.isArray(chunks) ? chunks : [];
    const lanes = Array.from({ length: count }, (_, laneIndex) => ({
        laneIndex,
        staged: items.some((item) => referenceLaneIndex(item) === laneIndex),
        resolved: 0,
        causeCounts: {
            [REFERENCE_LANE_CAUSE.BELOW_THRESHOLD]: 0,
            [REFERENCE_LANE_CAUSE.OUTSIDE]: 0,
            [REFERENCE_LANE_CAUSE.EXCLUDED]: 0,
        },
        dominantCause: null,
    }));

    for (const chunk of windows) {
        const { winners, verdicts } = resolveReferenceVerdicts({
            ...options,
            windowStart: chunk?.start,
            windowEnd: chunk?.end,
        });
        const seen = lanes.map(() => new Set());
        verdicts.forEach((verdict, itemIndex) => {
            const laneIndex = referenceLaneIndex(items[itemIndex]);
            if (laneIndex < 0 || laneIndex >= count) return;
            seen[laneIndex].add(verdict);
        });
        for (const lane of lanes) {
            if (!lane.staged) continue;
            if (winners[lane.laneIndex]) { lane.resolved += 1; continue; }
            const reasons = seen[lane.laneIndex];
            const cause = reasons.has(REFERENCE_VERDICT.BELOW_THRESHOLD)
                ? REFERENCE_LANE_CAUSE.BELOW_THRESHOLD
                : reasons.has(REFERENCE_VERDICT.OUTSIDE)
                    ? REFERENCE_LANE_CAUSE.OUTSIDE
                    : REFERENCE_LANE_CAUSE.EXCLUDED;
            lane.causeCounts[cause] += 1;
        }
    }
    for (const lane of lanes) {
        const ranked = [
            REFERENCE_LANE_CAUSE.BELOW_THRESHOLD,
            REFERENCE_LANE_CAUSE.OUTSIDE,
            REFERENCE_LANE_CAUSE.EXCLUDED,
        ].filter((cause) => lane.causeCounts[cause] > 0);
        lane.dominantCause = ranked[0] || null;
    }
    return lanes;
}

/**
 * One most-specific overlapping item per Reference lane, or null.
 * Mirrored in server/reference_resolution.py.
 */
export function resolveEffectiveReferences(options = {}) {
    // The extra `itemIndex` the core carries is internal bookkeeping; the
    // published winner shape stays `{laneIndex, item}`.
    return resolveReferenceVerdicts(options).winners.map(
        (winner) => (winner ? { laneIndex: winner.laneIndex, item: winner.item } : null),
    );
}

File: server/reference_resolution.py
"""Pure effective-Reference resolution shared conceptually with the frontend."""

from __future__ import annotations


# Type-homogeneous Bridge output groups. Mirrored in
# web/js/reference_resolution.js. Numbered socket names are presentation detail;
# recipes declare which whole bridge block they drive.
REFERENCE_OUTPUT_NAMES = (
    "image_slots",
    "audio_slots",
    "reference_prompt",
    "reference_names",
)

def reference_live_outputs(hard) -> set:
    """Return the outputs a recipe actually drives.

    A recipe without a `live_outputs` declaration (detached/custom) keeps every
    output live; anything else is dead and emits its documented fallback.
    """
    declared = hard.get("live_outputs") if isinstance(hard, dict) else None
    if not isinstance(declared, list):
        return set(REFERENCE_OUTPUT_NAMES)
    live = {str(name) for name in declared if isinstance(name, str)}
    return live.intersection(REFERENCE_OUTPUT_NAMES)


def _integer(value, default: int = 0) -> int:
    try:
        return int(value)
    except (TypeError, ValueError, OverflowError):
        return default


REFERENCE_VERDICT = {
    "WINNER": "winner", "SUPERSEDED": "superseded",
    "BELOW_THRESHOLD": "below_threshold", "OUTSIDE": "outside", "EXCLUDED": "excluded",
}
REFERENCE_VERDICT_LABEL = {
    "winner": "In window", "superseded": "Superseded",
    "below_threshold": "Below threshold", "outside": "Outside window", "excluded": "Excluded",
}


def _item_field(item, key, default=None):
    return item.get(key, default) if isinstance(item, dict) else getattr(item, key, default)


def resolve_reference_verdicts(
    *,
    reference_items,
    lane_count,
    scene_duration,
    window_start,
    window_end,
    lane_configs=None,
    frame_threshold_pct=0.0,
) -> dict:
    """Return one most-specific overlapping item per Reference lane.

    Ranges are half-open. A negative item end resolves to scene end before
    scoring. Muted items and hidden lanes do not participate.

    `frame_threshold_pct` drops an item whose in-window overlap is under that
    percentage of the shorter of its span and the window span, before the
    survivors are scored by overlap / item span. Containment in either
    direction always clears the threshold. Unlike the
    prompt boundary threshold there is no never-empty guard: References resolve
    to a single winner per lane, so a lane may legitimately resolve to nothing
    and report `has_reference = 0` for that window. That is the point of the
    setting, and it is why enqueue warns when it flips across a batch.

    Enqueue announces BOTH causes of that flip, under separate notifications:
    the threshold dropping a chunk the item does overlap, and the item simply
    not reaching a chunk at all. Only the first is a setting the user probably
    did not mean to hit, but both change which sockets are wired and so both
    change the inferred task mode. The frontend keeps the reason — see
    `classifyReferenceChunks` in `web/js/reference_resolution.js`; this module
    mirrors that verdict core so diagnostics use the same reasons.
    """
    count = max(1, _integer(lane_count, 1))
    duration = max(0, _integer(scene_duration, 0))
    start = min(duration, max(0, _integer(window_start, 0)))
    end = min(duration, max(start, _integer(window_end, duration)))
    configs = lane_configs if isinstance(lane_configs, list) else []
    items = reference_items if isinstance(reference_items, list) else []
    try:
        threshold = max(0.0, min(100.0, float(frame_threshold_pct or 0.0))) / 100.0
    except (TypeError, ValueError):
        threshold = 0.0
    winners = [None] * count
    scores = [None] * count
    winner_indices = [None] * count
    verdicts = {}

    for item_index, item in enumerate(items):
        get = lambda key, default=None: _item_field(item, key, default)
        lane_index = _integer(get("lane_index", 0), 0)
        if lane_index < 0 or lane_index >= count or bool(get("muted", False)):
            verdicts[item_index] = REFERENCE_VERDICT["EXCLUDED"]
            continue
        if lane_index < len(configs):
            config = configs[lane_index]
            hidden = config.get("hidden", False) if isinstance(config, dict) else getattr(config, "hidden", False)
            if bool(hidden):
                verdicts[item_index] = REFERENCE_VERDICT["EXCLUDED"]
                continue
        item_start = min(max(0, _integer(get("start_frame", 0), 0)), max(0, duration - 1))
        raw_end = _integer(get("end_frame", -1), -1)
        item_end = duration if raw_end < 0 else min(duration, raw_end)
        if item_end <= item_start:
            verdicts[item_index] = REFERENCE_VERDICT["OUTSIDE"]
            continue
        overlap = max(0, min(item_end, end) - max(item_start, start))
        if overlap <= 0:
            verdicts[item_index] = REFERENCE_VERDICT["OUTSIDE"]
            continue
        specificity = overlap / (item_end - item_start)
        coverage = overlap / max(1, min(item_end - item_start, end - start))
        if threshold > 0 and coverage < threshold:
            verdicts[item_index] = REFERENCE_VERDICT["BELOW_THRESHOLD"]
            continue
        score = (specificity, item_start, item_index)
        if scores[lane_index] is None or score > scores[lane_index]:
            if winner_indices[lane_index] is not None:
                verdicts[winner_indices[lane_index]] = REFERENCE_VERDICT["SUPERSEDED"]
            winner_indices[lane_index] = item_index
            verdicts[item_index] = REFERENCE_VERDICT["WINNER"]
            scores[lane_index] = score
            winners[lane_index] = {"laneIndex": lane_index, "item": item}
        else:
            verdicts[item_index] = REFERENCE_VERDICT["SUPERSEDED"]
    return {"winners": winners, "verdicts": verdicts}


def resolve_effective_references(**kwargs) -> list:
    """Preserve the published winner shape while deriving it from verdicts."""
    return resolve_reference_verdicts(**kwargs)["winners"]


def resolve_reference_staging(**kwargs) -> dict:
    """Index all authored staging, including losers; absence identifies deletion."""
    resolved = resolve_reference_verdicts(**kwargs)
    items, members = {}, {}
    for index, item in enumerate(kwargs.get("reference_items") or []):
        item_id = str(_item_field(item, "reference_item_id", "") or "")
        if not item_id:
            continue
        items[item_id] = {
            "verdict": resolved["verdicts"][index],
            "lane_index": _integer(_item_field(item, "lane_index", 0)),
            "start_frame": _item_field(item, "start_frame", 0),
            "end_frame": _item_field(item, "end_frame", -1),
        }
        for binding in _item_field(item, "members", []) or []:
            member_id = str(_item_field(binding, "member_id", "") or "")
            if member_id:
                members.setdefault(member_id, []).append(item_id)
    return {"resolved": True, "items": items, "members": members}
